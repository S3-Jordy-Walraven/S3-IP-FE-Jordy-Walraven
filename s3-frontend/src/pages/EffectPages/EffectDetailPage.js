import React, { useState, useEffect } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import Canvas from '../../components/EffectComponents/Canvas';
import EffectService from '../../services/EffectService';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';

const EffectDetailPage = (props) => {
    const { id } = useParams();
    const service = new EffectService();
    const [effect, setEffect] = useState(null);

    const [show, setShow] = useState(false);

    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);
    const navigate = useNavigate();

    useEffect(() => {
        service.getEffectById(id).then((data) => {
            console.log(data);
            setEffect(data.data);
        })
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    function backButtonCallback() {
        navigate("/")
    }

    async function downloadButtonCallback() {
        var anchor = document.createElement("a");
        console.log(effect.effectContent)
        anchor.href = 'data:attachment/text,' + await encodeURIComponent(effect.effectContent);
        anchor.target = '_blank';
        anchor.download = `${effect.effectName}.html`;
        anchor.click();
    }


    return (
        <div>
            <button onClick={backButtonCallback} className="btn btn-primary m-5" style={{ width: "5%", height: "5%" }}>Back</button>
            {
                effect !== null ?
                    <div>
                        <div style={{ textAlign: "center", marginTop: "2%", marginBottom: "-2%" }}>
                            <h1>{effect.effectName}</h1>

                            <Canvas finalString={effect != null ? effect.effectContent : null} />
                            <h2>By {effect.creatorName}</h2>
                        </div>
                        <div className='d-flex justify-content-center' style={{ marginTop: "3%" }}>
                            <button className='btn btn-primary m-2 p-3' onClick={handleShow}>View code</button>
                            <button className='btn btn-success m-2 p-3' onClick={downloadButtonCallback}>Download</button>
                        </div>

                        <Modal show={show} onHide={handleClose} size="xl">
                            <Modal.Header closeButton>
                                <Modal.Title>Effect Code</Modal.Title>
                            </Modal.Header>
                            <Modal.Body scrollable>{<pre><code>{effect.effectContent}</code></pre>}</Modal.Body>
                            <Modal.Footer>
                                <Button variant="secondary" onClick={handleClose}>
                                    Close
                                </Button>

                            </Modal.Footer>
                        </Modal>
                    </div>
                    : <>No effect with this id</>
            }
        </div>
    )
}

export default EffectDetailPage