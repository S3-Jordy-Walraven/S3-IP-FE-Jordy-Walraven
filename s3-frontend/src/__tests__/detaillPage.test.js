import { render, screen, cleanup, waitFor, } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from "react-router-dom";
import EffectDetailPage from "../pages/EffectPages/EffectDetailPage";

afterEach(cleanup);

test("should render without effect", () => {
    render(<Router><EffectDetailPage /></Router>);
    const effectUploadPageElement = screen.getByTestId("effectDetailPage-1");
    expect(effectUploadPageElement).toBeInTheDocument();
    expect(effectUploadPageElement).toHaveTextContent("No effect with this id");
});

test("should render effect", () => {
    render(<Router><EffectDetailPage effect={{ effectName: "TestEffect", creatorName: "Tester", effectContent: "<script>console.log('hello')</script>" }} /></Router>);

    const effectUploadPageElement = screen.getByTestId("effectDetailPage-1");
    expect(effectUploadPageElement).toBeInTheDocument();
    expect(effectUploadPageElement).toHaveTextContent("TestEffect");
    expect(effectUploadPageElement).toHaveTextContent("Tester");
    expect(effectUploadPageElement).toHaveTextContent("View code");
});


test("should view code", async() => {
    render(<Router><EffectDetailPage effect={{ effectName: "TestEffect", creatorName: "Tester", effectContent: "<script>console.log('hello')</script>" }} /></Router>);
    const viewCodeButton = screen.getByTestId("viewCodeButton-1");
    userEvent.click(viewCodeButton);
    const codeViewer = screen.getByTestId("codeViewer-1");
    await waitFor(()=>expect(codeViewer).toHaveTextContent("<script>console.log('hello')</script>"));
});

test("should run download button", async() => {
    render(<Router><EffectDetailPage effect={{ effectName: "TestEffect", creatorName: "Tester", effectContent: "<script>console.log('hello')</script>" }} /></Router>);
    const viewCodeButton = screen.getByTestId("downloadButton-1");
    userEvent.click(viewCodeButton);
});

test("should run oneClick button", async() => {
    render(<Router><EffectDetailPage effect={{ effectName: "TestEffect", creatorName: "Tester", effectContent: "<script>console.log('hello')</script>" }} /></Router>);
    const viewCodeButton = screen.getByTestId("onClick-1");
    userEvent.click(viewCodeButton);
});

test("should go back to home", async() => {
    render(<Router><EffectDetailPage effect={{ effectName: "TestEffect", creatorName: "Tester", effectContent: "<script>console.log('hello')</script>" }} /></Router>);
    const viewCodeButton = screen.getByTestId("backButton-1");
    userEvent.click(viewCodeButton);
});

test("matches snapshot", () => {
    const tree = renderer.create(<Router> <EffectDetailPage /></Router>).toJSON();
    expect(tree).toMatchSnapshot();
});
