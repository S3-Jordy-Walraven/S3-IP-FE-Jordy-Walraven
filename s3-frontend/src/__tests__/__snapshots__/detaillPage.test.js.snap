// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  " ",
  <div
    data-testid="effectDetailPage-1"
    style={
      Object {
        "color": "white",
      }
    }
  >
    <button
      className="btn btn-danger m-5 pt-3 pb-3"
      data-testid="backButton-1"
      onClick={[Function]}
    >
      Back to home
    </button>
    No effect with this id
  </div>,
]
`;
